### PR TITLE
[minor] Dedicated strided sum kernel

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Improves the FusedDropoutBias speed, turns out that `torch.sum()` can be suboptimal in that case (we're summing on the dimension which is not typically the contiguous one).

Tons of updated numbers, hard to show everything here, but basically fused dropout and fused linear layers are now faster across the board, sometimes much faster (the plots are in the changed files, all the new plots on a laptop 3080 -*previous plots were sometimes on a V100, hence a possible scale discrepancy, focus on the triton/pytorch gap-*). It's still kind of a mystery for me how to speed up the sum for very big dimensions (> 4096 pytorch becomes faster again), checking with @ptillet

Not too important because not life changing, but faster is always better I suppose.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: xformers/triton/dropout.py [hunk 1]
```diff
@@ -104,7 +104,20 @@ def grid(meta):
         # fmt: on
 
         if ctx.trainable_bias:
-            grad_bias: Optional[torch.Tensor] = torch.sum(grad_in, dim=0)
+            grad_bias: Optional[torch.Tensor] = torch.empty(
```
Comment (blefaudeux) [inline xformers/triton/dropout.py:107]: @jieru-hu turns out torch.sum() is not the fastest take out there

**Comment by blefaudeux** [comment]
Screen Shot 2021-11-12 at 4 20 37 PM
back to draft, looks like there's a scheduling issue (should be easy to fix). Benchmarks for the win !

**Comment by blefaudeux** [comment]
two examples of changes (mind the gap, left was on a V100 and right is on a laptop ampere, but the difference with pytorch is visible)

Screenshot from 2021-12-03 23-13-57
Screenshot from 2021-12-03 23-13-22

**Comment by blefaudeux** [comment]
Discussing with @Chillee on another channel, looks like this could be significantly sped up. Looks like we could also give NVFuser a go, it shows very good perf

**Comment by blefaudeux** [comment]
reopening for review @dianaml0 @fmassa, quickly testing the suggestions from Horace on another PR which would end up on top of this one. In particular if the scheduling is changed on the BW pass then the grad bias can be fused altogether with the input grad computations, which could be beneficial (this path seems to be very fast with nvfuser)

**Comment by blefaudeux** [comment]
minor bugfix, ready for review @dianaml0 @fmassa, should be relatively side effect-free. It's still not ideal for dropout, in that the BW pass can be completely fused in that case, I've worked on that on another branch which shows even better speed (basically up to bandwidth in the forward pass (20% better than this PR, give or take), and a bit better than this PR for the BW), but not completely ready

**Comment by blefaudeux** [comment]
Guessing that this is fine, small enough and no comments..

**Comment by dianaml0** [comment]
Sorry I couldn't finish reviewing this earlier @blefaudeux but interesting PR to go through! The comments really helped :)
